### PR TITLE
Broadcaster component

### PR DESCRIPTION
**Pull request by thibault-martinez**
This PR adds a working implementation for TCP and UDP broadcasting to a node tethered neighbours.

# Test Plan:
The main function currently act as a simple node repeatedly broadcasting the same hash to a TCP neighbour and a UDP neighbour.

It can be tested by connecting two nc clients to it 

`nc 127.0.0.1 14260 -p 14262` for TCP

`echo | nc -u 127.0.0.1 14261 -p 14263` for UDP

### CODE REVIEW

File: ciri/main.c [hunk 1]
```diff
@@ -25,7 +25,20 @@ int main() {
...
+  // TODO(thibault) Remove, for broadcaster testing purpose
+  neighbor_t neighbor;
+  neighbor_init_with_uri(&neighbor, "tcp://127.0.0.1:14262");
+  neighbor_add(node_g.neighbors, neighbor);
+  neighbor_init_with_uri(&neighbor, "udp://127.0.0.1:14263");
+  neighbor_add(node_g.neighbors, neighbor);
+  flex_trit_t raw_trits[] = {-1, 0, 1, 1, 0, 1, 1,  0, -1,
+                             -1, 1, 0, 1, 1, 0, -1, 1};
+  while (1) {
+    trit_array_p trits = trit_array_new(17);
```
Comment (th0br0) [inline ciri/main.c:37]: this is a memleak, could you fix? (even though this is just for debugging

File: common/network/lists/concurrent_list.c.inc [hunk 2]
```diff
@@ -149,12 +150,12 @@
...
   concurrent_list_status_t CONCURRENT_LIST_OF_##TYPE##_PUSH_BACK( \
       concurrent_list_of_##TYPE *const self, TYPE const data) {   \
-    concurrent_list_node_of_##TYPE##_t *node;                     \
+    concurrent_list_node_of_##TYPE *node;                         \
     if (self == NULL) {                                           \
       return CONCURRENT_LIST_NULL_SELF;                           \
     }                                                             \
-    node = (concurrent_list_node_of_##TYPE##_t *)calloc(          \
-        1, sizeof(concurrent_list_node_of_##TYPE##_t));           \
+    node = (concurrent_list_node_of_##TYPE *)calloc(              \
```
Comment (th0br0) [inline common/network/lists/concurrent_list.c.inc]: Could you open an issue to implement a statically-sized list that can be used as a drop-in for this?
Not needed right now, but for some targets in the future it'll probably be

File: common/network/services/tcp_sender.cc [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+#include <boost/asio.hpp>
+
+#include "common/network/logger.h"
+#include "common/network/services/tcp_sender.hpp"
+
+bool tcp_send(endpoint_t *const endpoint, trit_array_p const hash) {
+  if (endpoint == NULL || endpoint->opaque_inetaddr == NULL) {
+    return false;
+  }
+  size_t trytes_num = num_trytes_for_trits(hash->num_trits);
```
Comment (th0br0) [inline common/network/services/tcp_sender.cc:17]: We have some macros for this?

File: common/network/services/udp_sender.cc [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+#include "common/network/services/udp_sender.hpp"
+
+bool udp_send(endpoint_t *const endpoint, trit_array_p const hash) {
+  boost::asio::ip::address addr;
+
+  if (endpoint == NULL || endpoint->opaque_inetaddr == NULL) {
+    return false;
+  }
+  size_t trytes_num = num_trytes_for_trits(hash->num_trits);
+  trit_t trytes[trytes_num];
```
Comment (th0br0) [inline common/network/services/udp_sender.cc]: We have a macro for this?

**Review by th0br0 — CHANGES_REQUESTED** [review]

File: ciri/main.c (hunk 1 above)
Comment (thibault-martinez) [reply ciri/main.c:37]: It is freed at the end of the chain, in the broadcaster (https://github.com/iotaledger/entangled/pull/117/files#diff-855c1ef07801182e60e316606f5cadceR27)

File: common/network/services/tcp_sender.cc (hunk 3 above)
Comment (thibault-martinez) [reply common/network/services/tcp_sender.cc:17]: Couldn't find it, do we ?

File: common/network/services/udp_sender.cc (hunk 4 above)
Comment (thibault-martinez) [reply common/network/services/udp_sender.cc]: ditto

Comment (th0br0) [reply common/network/services/udp_sender.cc]: I misread, you're right.

However the type is wrong should be `tryte_t trytes`

File: common/network/services/tcp_sender.cc [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  trit_t trytes[trytes_num];
```
Comment (th0br0) [inline common/network/services/tcp_sender.cc]: wrong type

File: ciri/main.c (hunk 1 above)
Comment (th0br0) [reply ciri/main.c:37]: `trit_array_new` is only available if dynamic allocation is available though :thinking:

Comment (th0br0) [reply ciri/main.c:37]: Anyway, fine for now then.

**Review by th0br0 — CHANGES_REQUESTED** [review]

**Review by th0br0 — APPROVED** [review]

**Pull request merged**